### PR TITLE
Add debugging steps for DiscoverEC2 User Task issues

### PR DESCRIPTION
**Pull request by marcoandredinis**
This PR adds a new field when getting an User Task. Description contains a markdown document that should be displayed to the user which helps them fix the issue with the task.

### CODE REVIEW

**Review by hugoShaka — COMMENTED** [review]
The code LGTM, my only question is regarding the description location. Do we want to hardcode those in `api/` or should the description be set server-side and live in `lib/`?

**Comment by marcoandredinis** [comment]
> The code LGTM, my only question is regarding the description location. Do we want to hardcode those in `api/` or should the description be set server-side and live in `lib/`?

That's a good point. I'm not sure which one is a better fit.
IMO, it's fine in both places.
If you or someone thinks this should be placed under `lib/` I'm fine changing it.

File: api/types/usertasks/descriptions/ec2-ssm-agent-connection-lost.md
Comment (r0mant) [inline api/types/usertasks/descriptions/ec2-ssm-agent-connection-lost.md:2]:
```suggestion
Some instances appear to have lost connection to Amazon Systems Manager.
```

File: api/types/usertasks/descriptions/ec2-ssm-agent-connection-lost.md
Comment (r0mant) [inline api/types/usertasks/descriptions/ec2-ssm-agent-connection-lost.md:17]:
```suggestion
```

File: api/types/usertasks/descriptions/ec2-ssm-agent-connection-lost.md
Comment (r0mant) [inline api/types/usertasks/descriptions/ec2-ssm-agent-connection-lost.md:19]:
```suggestion
**Instance is missing IAM policy**
```

File: api/types/usertasks/descriptions/ec2-ssm-agent-connection-lost.md
Comment (r0mant) [inline api/types/usertasks/descriptions/ec2-ssm-agent-connection-lost.md:24]:
```suggestion
```

File: api/types/usertasks/descriptions/ec2-ssm-agent-not-registered.md
Comment (r0mant) [inline api/types/usertasks/descriptions/ec2-ssm-agent-not-registered.md:2]:
```suggestion
Some instances failed to connect to Amazon Systems Manager.
```

File: api/types/usertasks/descriptions/ec2-ssm-invocation-failure.md
Comment (r0mant) [inline api/types/usertasks/descriptions/ec2-ssm-invocation-failure.md:2]:
```suggestion
Some instances failed to communicate with the AWS Systems Manager service to execute the install script.
```

File: api/types/usertasks/descriptions/ec2-ssm-invocation-failure.md
Comment (r0mant) [inline api/types/usertasks/descriptions/ec2-ssm-invocation-failure.md:18]: How/where can a user find this document?

File: api/types/usertasks/descriptions/ec2-ssm-script-failure.md
Comment (r0mant) [inline api/types/usertasks/descriptions/ec2-ssm-script-failure.md:1]: Also, how can a user see and update the script?
```suggestion
Teleport was able to reach the SSM Agent inside the EC2 instance, however the install script returned an error.
```

File: api/types/usertasks/descriptions/ec2-ssm-script-failure.md
Comment (r0mant) [inline api/types/usertasks/descriptions/ec2-ssm-script-failure.md:3]: The URL seems to be missing?

File: api/types/usertasks/descriptions/ec2-ssm-unsupported-os.md
Comment (r0mant) [inline api/types/usertasks/descriptions/ec2-ssm-unsupported-os.md:9]:
```suggestion
```

**Review by r0mant — COMMENTED** [review]
With some comments.

File: api/types/usertasks/descriptions/ec2-ssm-script-failure.md
Comment (marcoandredinis) [reply api/types/usertasks/descriptions/ec2-ssm-script-failure.md:3]: The URL will appear below.
We have the the description and then a table with affected instances
That table will contain, for each instance, its invocation URL.

**Review by hugoShaka — APPROVED** [review]
LGTM

I'd prefer the description to be set server-side and live `lib/` but that's not blocking.

**Comment by marcoandredinis** [comment]
> LGTM
> 
> I'd prefer the description to be set server-side and live `lib/` but that's not blocking.

Done 👍

File: api/types/usertasks/descriptions/ec2-ssm-invocation-failure.md
Comment (marcoandredinis) [reply api/types/usertasks/descriptions/ec2-ssm-invocation-failure.md]: I'll the SSM Document name to the list of fields.
UI should then build the SSM Document URL (we have the region and the SSM Document name, so that should be simple).

File: api/types/usertasks/descriptions/ec2-ssm-script-failure.md
Comment (marcoandredinis) [reply api/types/usertasks/descriptions/ec2-ssm-script-failure.md]: I'll the InstallerScript used to the list of fields.
UI can suggest the user to do a `tctl edit installer/<script-name>` so that the user can edit the script.
What do you think?

I'm not so sure about this one as I am for the SSM Document, because for this one, the user needs to change tools (we don't expose CRUD for installer scripts in the UI).

I'll add them anyway.

**Comment by marcoandredinis** [comment]
@r0mant I've added two new fields: SSM Document and Installer Script

We can show them in the UI.
For the SSM Document we can link out to Amazon Console (we have region and the document name, and that's all it's needed to build the URL)
For the Installer Script we can teach the user how to do it with `tctl` for now (while we don't expose that resource in the UI).

Can you please take another look?

**Review by kimlisa — APPROVED** [review]

**Comment by kimlisa** [comment]
> @kimlisa I know you were looking for a way to render this markdown in the UI. Did you reach any conclusion? Can we pursue this method or should we think of some other way to show those instructions?

yeah, assist uses `react-markdown` and we'll do the same

File: api/types/usertasks/descriptions/ec2-ssm-invocation-failure.md
Comment (r0mant) [reply api/types/usertasks/descriptions/ec2-ssm-invocation-failure.md]: @marcoandredinis I think these messages should be templated using text/template. Then whatever returns these messages to the web UI can parameterize them with appropriate data without having to rely on the web UI to assemble a bunch of stuff.

File: lib/usertasks/descriptions/ec2-ssm-invocation-failure.md
Comment (r0mant) [inline lib/usertasks/descriptions/ec2-ssm-invocation-failure.md:19]: It is still not clear where this SSM document or install script is. I think we need to template these messages like in my other comment to make them more useful to the user.

**Review by r0mant — COMMENTED** [review]
@marcoandredinis @kimlisa Before merging this, I'd like to see how these are presented in the web UI so we can see if things make sense or need more tweaks to be more clear and actionable. I understand this depends on having this view implemented in the web UI so I think it's ok to hold this until we have it.

File: api/types/usertasks/descriptions/ec2-ssm-invocation-failure.md
Comment (marcoandredinis) [reply api/types/usertasks/descriptions/ec2-ssm-invocation-failure.md]: We can template everything or we show have specific fields for metadata and then only the description is templated.

Taking the following screen as example
image

I was thinking that only the `Details` section was to be converted from Markdown into HTML (aka sent over the API).

After reading your comment I wonder if you want it all to be generated in the backend.
Which parts should be rendered in the backend?
Title
Metadata (region, account id)
Details
Impacted instances

We can do it all, if that's what we want.
However, we have to be careful because there are some elements that might look out of place (using a different CSS style) because the library we'll be using will not use the same set of styles that we have.
While this might be easy to fix for things that are mostly text (bolds, lists and such), it might not be easy for more complex elements like tables.

(@kimlisa please correct me if I'm wrong because I don't know the library we'll be using)

Comment (michellescripts) [reply api/types/usertasks/descriptions/ec2-ssm-invocation-failure.md]: React-markdown sounds OK, but we don't actually use it right now since Assist is deprecated. 

Is there a reason to send markdown to the front end, vs sending an error code indicating which display to show, and the UI has a collection of views?

Comment (r0mant) [reply api/types/usertasks/descriptions/ec2-ssm-invocation-failure.md]: @michellescripts The reason I'd prefer for the frontend to just be a simpler renderer in this case is so we can easily expand our "library" of errors (with descriptions and other metadata) without having to update UI every time we want to add a new type of user task.

Comment (michellescripts) [reply api/types/usertasks/descriptions/ec2-ssm-invocation-failure.md]: I'm fine with that! I do agree that templating in links for the user will be a better UX but we can tackle that whenever.

Comment (marcoandredinis) [reply api/types/usertasks/descriptions/ec2-ssm-invocation-failure.md]: We have 4 sections: title, metadata, description and table with affected resources.

My suggestion:
- Title is just a string.
- Metadata should come from the backend and separate fields and rendered in the frontend.
- Description comes from the backend as markdown and is rendered in the frontend using react-markdown.
- Table comes as a list from the backend, and is rendered in the frontend.

**Comment by marcoandredinis** [comment]
@r0mant Can we merge this one?

I think we came to a conclusion in this discussion thread
https://github.com/gravitational/teleport/pull/47513#discussion_r1875834404

**Review by r0mant — APPROVED** [review]
@marcoandredinis Yeah let's get this in.

**Pull request merged**
